crates/autopilot/src/decoded_settlement.rs

```diff
@@ -280,91 +268,49 @@ impl DecodedSettlement {
-        mut orders: Vec<OrderExecution>,
-    ) -> U256 {
-        self.trades.iter().fold(0.into(), |acc, trade| {
-            match orders.iter().position(|order| {
-                trade.matches_execution(order, &self.tokens, &self.domain_separator)
-            }) {
-                Some(i) => {
-                    // It's possible to have multiple fills with the same `executed_amount` for the
-                    // same order with different `solver_fees`. To end up with the correct total
-                    // fees we can only use every `OrderExecution` exactly once.
-                    let order = orders.swap_remove(i);
-                    acc + match self.fee(external_prices, &order, trade) {
-                        Some(fees) => fees.native,
-                        None => {
-                            tracing::warn!("possible incomplete fee calculation");
-                            0.into()
-                        }
-                    }
-                }
-                None => {
-                    tracing::warn!(
-                        "order not found for trade, possible incomplete fee calculation"
-                    );
-                    acc
-                }
-            }
-        })
-    }
-
-    /// Returns the list of executions with their fees,
-    /// which are supposed to be updated whenever a new settlement is executed.
-    /// Done for the orders that have solver-computed fees (limit orders).
-    pub fn order_executions(
+    /// Returns unsubsidized fees for all trades.
+    /// Length of the returned vector is equal to the length of `self.trades`
+    /// and `order_fees`.
+    pub fn all_fees(
         &self,
         external_prices: &ExternalPrices,
-        mut orders: Vec<OrderExecution>,
+        order_fees: &[(OrderUid, Option<U256>)],
     ) -> Vec<Fees> {
         self.trades
             .iter()
-            .filter_map(|trade| {
-                let i = orders.iter().position(|order| {
-                    trade.matches_execution(order, &self.tokens, &self.domain_separator)
-                })?;
-                // It's possible to have multiple fills with the same `executed_amount` for
-                // the same order with different `solver_fees`. To
-                // end up with the correct total fees we can only
-                // use every `OrderExecution` exactly once.
-                let order = orders.swap_remove(i);
-
-                // Update fee only for orders where fee is taken from surplus
-                if !matches!(order.executed_fee, ExecutedFee::Surplus) {
-                    return None;
-                }
-
-                let fees = self.fee(external_prices, &order, trade);
-
-                if fees.is_none() {
-                    tracing::warn!("possible incomplete fee calculation");
+            .zip(order_fees.iter())
+            .map(|(trade, (order, order_fee))| {
+                match self.fee(external_prices, *order, *order_fee, trade) {
```

Comment:
nit: this looks like a use case for `unwrap_or_else`

Reply:
Indeed, adjusted.